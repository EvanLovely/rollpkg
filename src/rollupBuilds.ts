import { resolve } from 'path';
import * as fs from 'fs-extra';
import {
  rollup,
  watch,
  Plugin,
  OutputPlugin,
  TreeshakingOptions,
  RollupWatchOptions,
  RollupBuild,
  RollupOutput,
} from 'rollup';
import resolveRollup from '@rollup/plugin-node-resolve';
import commonjs from '@rollup/plugin-commonjs';
import json from '@rollup/plugin-json';
import rollupTypescript from 'rollup-plugin-typescript2';
import replace from '@rollup/plugin-replace';
import invariantPlugin from 'rollup-plugin-invariant';
import { terser } from 'rollup-plugin-terser';
// babel only used to add /*#__PURE__*/ annotations for function calls using babel-plugin-annotate-pure-calls
import babel from '@rollup/plugin-babel';
import { DEFAULT_EXTENSIONS } from '@babel/core';

import {
  clearLine,
  clearConsole,
  convertPkgNameToKebabCase,
  convertKebabCaseToPascalCase,
} from './utils';
import { logTsError, logError } from './errorUtils';

/////////////////////////////////////
// create rollup config
interface CreateRollupConfig {
  (inputs: {
    pkgName: string;
    pkgSideEffects: boolean;
    pkgPeerDependencies: string[];
  }): {
    pkgPeerDependencyGlobals: { [key: string]: string };
    pkgUmdName: string;
    buildPlugins: Plugin[];
    prodBuildPlugins: Plugin[];
    outputPlugins: OutputPlugin[];
    outputProdPlugins: OutputPlugin[];
    treeshakeOptions: TreeshakingOptions;
  };
}

export const createRollupConfig: CreateRollupConfig = ({
  pkgName,
  pkgSideEffects,
  pkgPeerDependencies,
}) => {
  const pkgPeerDependencyGlobals: { [key: string]: string } = {};
  pkgPeerDependencies.forEach((peerDep) => {
    pkgPeerDependencyGlobals[peerDep] = convertKebabCaseToPascalCase(
      convertPkgNameToKebabCase(peerDep),
    );
  });
  const pkgUmdName = convertKebabCaseToPascalCase(pkgName);

  const buildPlugins: Plugin[] = [
    resolveRollup(),
    commonjs(),
    json(),
    rollupTypescript({
      // verbosity: 3, // use to debug
      tsconfigOverride: {
        compilerOptions: {
          types: [],
          declaration: true,
          sourceMap: true,
          rootDir: './src',
        },
        include: ['src', './node_modules/rollpkg/configs/types'],
        exclude: [
          '**/__tests__',
          '**/__mocks__',
          '**/__snapshots__',
          '**/*.test.*',
          '**/*.spec.*',
          '**/*.mock.*',
        ],
      },
      include: ['**/*.ts+(|x)', '**/*.js+(|x)'],
    }),
    replace({ __DEV__: "process.env.NODE_ENV !== 'production'" }),
    // eslint-disable-next-line @typescript-eslint/ban-ts-comment
    // @ts-ignore invariantPlugin is missing name, this isn't causing an actual error AFAIK, remove when pr is released: https://github.com/apollographql/invariant-packages/pull/45
    invariantPlugin(),
  ];

  if (pkgSideEffects === false)
    buildPlugins.push(
      babel({
        babelHelpers: 'bundled',
        extensions: [...DEFAULT_EXTENSIONS, '.ts', '.tsx'],
        plugins: ['annotate-pure-calls'],
        skipPreflightCheck: true,
        configFile: false,
        babelrc: false,
      }),
    );

  const prodBuildPlugins: Plugin[] = [
    ...buildPlugins,
    replace({ 'process.env.NODE_ENV': JSON.stringify('production') }),
  ];

  const outputPlugins: OutputPlugin[] = [];

  const outputProdPlugins: OutputPlugin[] = [terser()];

  const treeshakeOptions: TreeshakingOptions = {
    annotations: true,
    moduleSideEffects: pkgSideEffects,
    propertyReadSideEffects: pkgSideEffects,
    tryCatchDeoptimization: pkgSideEffects,
    unknownGlobalSideEffects: pkgSideEffects,
  };

  return {
    pkgPeerDependencyGlobals,
    pkgUmdName,
    buildPlugins,
    prodBuildPlugins,
    outputPlugins,
    outputProdPlugins,
    treeshakeOptions,
  };
};
/////////////////////////////////////

/////////////////////////////////////
// rollup watch
interface RollupWatch {
  (input: {
    pkgName: string;
    pkgDependencies: string[];
    pkgPeerDependencies: string[];
    input: string;
    treeshakeOptions: TreeshakingOptions;
    buildPlugins: Plugin[];
    outputPlugins: OutputPlugin[];
  }): void;
}

export const rollupWatch: RollupWatch = ({
  pkgName,
  pkgDependencies,
  pkgPeerDependencies,
  input,
  treeshakeOptions,
  buildPlugins,
  outputPlugins,
}) => {
  process.on('SIGINT', function () {
    clearLine();
    console.log('ROLLPKG WATCH END 👀👋', '\n');
    process.exit(0);
  });

  const watchOptions: RollupWatchOptions = {
    external: [...pkgDependencies, ...pkgPeerDependencies],
    input,
    treeshake: treeshakeOptions,
    plugins: buildPlugins,
    output: [
      {
        file: `dist/${pkgName}.esm.js`,
        format: 'esm',
        sourcemap: true,
        plugins: outputPlugins,
      },
      {
        file: `dist/${pkgName}.cjs.js`,
        format: 'cjs',
        sourcemap: true,
        plugins: outputPlugins,
      },
    ],
  };

  const watcher = watch(watchOptions);

  const earth = ['🌎', '🌏', '🌍'];
  let currentEarth = 2;
  const rotateEarth = () => {
    if (currentEarth === 2) currentEarth = 0;
    else {
      currentEarth = currentEarth + 1;
    }
    return earth[currentEarth];
  };

  watcher.on('event', (event) => {
    switch (event.code) {
      case 'START':
        break;
      case 'BUNDLE_START':
        clearConsole();
        console.log(`${rotateEarth()} Rollpkg building...`, '\n');
        break;
      case 'BUNDLE_END':
        clearConsole();
        console.log(`${earth[currentEarth]} Rollpkg build successful!`, '\n');
        console.log('Watching for changes...', '\n');
        break;
      case 'END':
        break;
      case 'ERROR':
        clearConsole();
        if (event.error.plugin === 'rpt2') {
          logTsError({ message: event.error.message });
        } else {
          logError({ fullError: event.error });
        }
        console.log('Watching for changes...', '\n');
        break;
    }
  });
};
/////////////////////////////////////

/////////////////////////////////////
// create rollup bundles
interface BuildBundles {
  (input: {
    input: string;
    pkgDependencies: string[];
    pkgPeerDependencies: string[];
    treeshakeOptions: TreeshakingOptions;
    buildPlugins: Plugin[];
    prodBuildPlugins: Plugin[];
  }): Promise<[RollupBuild, RollupBuild, RollupBuild, RollupBuild]>;
}

export const buildBundles: BuildBundles = ({
  input,
  pkgDependencies,
  pkgPeerDependencies,
  treeshakeOptions,
  buildPlugins,
  prodBuildPlugins,
}) =>
  Promise.all([
    // esm and cjs development builds
    rollup({
      external: [...pkgDependencies, ...pkgPeerDependencies],
      input,
      treeshake: treeshakeOptions,
      plugins: buildPlugins,
    }),

    // cjs production build
    rollup({
      external: [...pkgDependencies, ...pkgPeerDependencies],
      input,
      treeshake: treeshakeOptions,
      plugins: prodBuildPlugins,
    }),

    // umd development build
    rollup({
      external: [...pkgPeerDependencies],
      input,
      treeshake: treeshakeOptions,
      plugins: buildPlugins,
    }),

    // umd production build
    rollup({
      external: [...pkgPeerDependencies],
      input,
      treeshake: treeshakeOptions,
      plugins: prodBuildPlugins,
    }),
  ]);
/////////////////////////////////////

/////////////////////////////////////
// write rollup bundles
interface WriteBundles {
  (input: {
    pkgName: string;
    bundle: RollupBuild;
    bundleProd: RollupBuild;
    bundleUmd: RollupBuild;
    bundleUmdProd: RollupBuild;
    outputPlugins: OutputPlugin[];
    outputProdPlugins: OutputPlugin[];
    pkgUmdName: string;
    pkgPeerDependencyGlobals: { [key: string]: string };
  }): Promise<
    [RollupOutput, RollupOutput, RollupOutput, RollupOutput, RollupOutput, void]
  >;
}

export const writeBundles: WriteBundles = ({
  pkgName,
  bundle,
  bundleProd,
  bundleUmd,
  bundleUmdProd,
  outputPlugins,
  outputProdPlugins,
  pkgUmdName,
  pkgPeerDependencyGlobals,
}) => {
  // prettier-ignore
  const cjsEntryContent = 
`'use strict';

if (process.env.NODE_ENV === 'production') {
  module.exports = require('./${pkgName}.cjs.production.js');
} else {
  module.exports = require('./${pkgName}.cjs.development.js');
}`;

  return Promise.all([
    bundle.write({
      file: `dist/${pkgName}.esm.js`,
      format: 'esm',
      sourcemap: true,
      plugins: outputPlugins,
    }),

    bundle.write({
      file: `dist/${pkgName}.cjs.development.js`,
      format: 'cjs',
      sourcemap: true,
      plugins: outputPlugins,
    }),

    bundleProd.write({
      file: `dist/${pkgName}.cjs.production.js`,
      format: 'cjs',
      sourcemap: true,
      plugins: outputProdPlugins,
    }),

    bundleUmd.write({
      file: `dist/${pkgName}.umd.development.js`,
      format: 'umd',
      name: pkgUmdName,
      globals: pkgPeerDependencyGlobals,
      sourcemap: true,
      plugins: outputPlugins,
    }),

    bundleUmdProd.write({
      file: `dist/${pkgName}.umd.production.js`,
      format: 'umd',
      name: pkgUmdName,
      globals: pkgPeerDependencyGlobals,
      sourcemap: true,
      plugins: outputProdPlugins,
    }),
    fs.writeFile(
      resolve(process.cwd(), 'dist', `${pkgName}.cjs.js`),
      cjsEntryContent,
      'utf8',
    ),
  ]);
};
/////////////////////////////////////
