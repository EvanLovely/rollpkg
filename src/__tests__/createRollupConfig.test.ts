import { createRollupConfig } from '../rollupBuilds';

describe('treeshake options', () => {
  test('treeshake options with sideEffects: false', () => {
    const { treeshakeOptions } = createRollupConfig({
      addUmdBuild: false,
      kebabCasePkgName: 'test-package-name',
      pkgJsonSideEffects: false,
      pkgJsonPeerDependencyKeys: [],
    });

    expect(treeshakeOptions).toMatchInlineSnapshot(`
      Object {
        "annotations": true,
        "moduleSideEffects": false,
        "propertyReadSideEffects": false,
        "tryCatchDeoptimization": false,
        "unknownGlobalSideEffects": false,
      }
    `);
  });

  test('treeshake options with sideEffects: true', () => {
    const { treeshakeOptions } = createRollupConfig({
      addUmdBuild: false,
      kebabCasePkgName: 'test-package-name',
      pkgJsonSideEffects: true,
      pkgJsonPeerDependencyKeys: [],
    });

    expect(treeshakeOptions).toMatchInlineSnapshot(`
      Object {
        "annotations": true,
        "moduleSideEffects": true,
        "propertyReadSideEffects": true,
        "tryCatchDeoptimization": true,
        "unknownGlobalSideEffects": true,
      }
    `);
  });
});

describe('replaces dev mode code', () => {
  test("default build plugins replaces __DEV__ with process.env.NODE_ENV !== 'production'", () => {
    const { buildPluginsDefault } = createRollupConfig({
      addUmdBuild: false,
      kebabCasePkgName: 'test-package-name',
      pkgJsonSideEffects: true,
      pkgJsonPeerDependencyKeys: [],
    });

    expect(buildPluginsDefault).toContainEqual({
      name: '@rollup/plugin-replace',
      calledWith: {
        __DEV__: "process.env.NODE_ENV !== 'production'",
      },
    });
  });

  test("dev build plugins replaces NODE_ENV with 'development'", () => {
    const { buildPluginsWithNodeEnvDevelopment } = createRollupConfig({
      addUmdBuild: false,
      kebabCasePkgName: 'test-package-name',
      pkgJsonSideEffects: true,
      pkgJsonPeerDependencyKeys: [],
    });

    expect(buildPluginsWithNodeEnvDevelopment).toContainEqual({
      name: '@rollup/plugin-replace',
      calledWith: { 'process.env.NODE_ENV': JSON.stringify('development') },
    });
  });

  test("prod build plugins replaces NODE_ENV with 'production'", () => {
    const { buildPluginsWithNodeEnvProduction } = createRollupConfig({
      addUmdBuild: false,
      kebabCasePkgName: 'test-package-name',
      pkgJsonSideEffects: true,
      pkgJsonPeerDependencyKeys: [],
    });

    expect(buildPluginsWithNodeEnvProduction).toContainEqual({
      name: '@rollup/plugin-replace',
      calledWith: { 'process.env.NODE_ENV': JSON.stringify('production') },
    });
  });
});

describe('umd build configuration', () => {
  test('umd package name', () => {
    const { umdNameForPkg } = createRollupConfig({
      addUmdBuild: true,
      kebabCasePkgName: 'test-package-name',
      pkgJsonSideEffects: true,
      pkgJsonPeerDependencyKeys: ['some-peer-dependency'],
    });

    expect(umdNameForPkg).toBe('TestPackageName');
  });

  test('umd external globals when umdGlobalDependencies is not specified', () => {
    const {
      umdExternalDependencies,
      umdDependencyGlobals,
    } = createRollupConfig({
      addUmdBuild: true,
      kebabCasePkgName: 'test-package-name',
      pkgJsonSideEffects: true,
      pkgJsonPeerDependencyKeys: ['some-peer-dependency'],
    });

    expect(umdExternalDependencies).toContain('some-peer-dependency');
    expect(umdDependencyGlobals).toMatchObject({
      'some-peer-dependency': 'SomePeerDependency',
    });
  });

  test('umd external globals when umdGlobalDependencies is specified', () => {
    const {
      umdExternalDependencies,
      umdDependencyGlobals,
    } = createRollupConfig({
      addUmdBuild: true,
      kebabCasePkgName: 'test-package-name',
      pkgJsonSideEffects: true,
      pkgJsonPeerDependencyKeys: ['some-peer-dependency'],
      pkgJsonUmdGlobalDependencies: {
        'some-listed-umd-global-dependency': 'TheGlobalTestDependency',
      },
    });

    expect(umdExternalDependencies).toContain(
      'some-listed-umd-global-dependency',
    );
    expect(umdDependencyGlobals).toMatchObject({
      'some-listed-umd-global-dependency': 'TheGlobalTestDependency',
    });
  });
});

describe('rollup plugins', () => {
  test('default build plugins with sideEffects: false', () => {
    const { buildPluginsDefault } = createRollupConfig({
      addUmdBuild: true,
      kebabCasePkgName: 'test-package-name',
      pkgJsonSideEffects: false,
      pkgJsonPeerDependencyKeys: [],
    });

    expect(buildPluginsDefault).toMatchSnapshot();
  });

  test('default build plugins with sideEffects: true', () => {
    const { buildPluginsDefault } = createRollupConfig({
      addUmdBuild: false,
      kebabCasePkgName: 'test-package-name',
      pkgJsonSideEffects: true,
      pkgJsonPeerDependencyKeys: [],
    });

    expect(buildPluginsDefault).toMatchSnapshot();
  });

  test('dev build plugins with sideEffects: false', () => {
    const { buildPluginsWithNodeEnvDevelopment } = createRollupConfig({
      addUmdBuild: false,
      kebabCasePkgName: 'test-package-name',
      pkgJsonSideEffects: false,
      pkgJsonPeerDependencyKeys: [],
    });

    expect(buildPluginsWithNodeEnvDevelopment).toMatchSnapshot();
  });

  test('dev build plugins with sideEffects: true', () => {
    const { buildPluginsWithNodeEnvDevelopment } = createRollupConfig({
      addUmdBuild: false,
      kebabCasePkgName: 'test-package-name',
      pkgJsonSideEffects: true,
      pkgJsonPeerDependencyKeys: [],
    });

    expect(buildPluginsWithNodeEnvDevelopment).toMatchSnapshot();
  });

  test('prod build plugins with sideEffects: false', () => {
    const { buildPluginsWithNodeEnvProduction } = createRollupConfig({
      addUmdBuild: false,
      kebabCasePkgName: 'test-package-name',
      pkgJsonSideEffects: false,
      pkgJsonPeerDependencyKeys: [],
    });

    expect(buildPluginsWithNodeEnvProduction).toMatchSnapshot();
  });

  test('prod build plugins with sideEffects: true', () => {
    const { buildPluginsWithNodeEnvProduction } = createRollupConfig({
      addUmdBuild: false,
      kebabCasePkgName: 'test-package-name',
      pkgJsonSideEffects: true,
      pkgJsonPeerDependencyKeys: [],
    });

    expect(buildPluginsWithNodeEnvProduction).toMatchSnapshot();
  });

  test('output default plugins', () => {
    const { outputPluginsDefault } = createRollupConfig({
      addUmdBuild: false,
      kebabCasePkgName: 'test-package-name',
      pkgJsonSideEffects: false,
      pkgJsonPeerDependencyKeys: [],
    });

    expect(outputPluginsDefault).toMatchSnapshot();
  });

  test('output prod plugins', () => {
    const { outputPluginsProduction } = createRollupConfig({
      addUmdBuild: false,
      kebabCasePkgName: 'test-package-name',
      pkgJsonSideEffects: false,
      pkgJsonPeerDependencyKeys: [],
    });

    expect(outputPluginsProduction).toMatchSnapshot();
  });
});
