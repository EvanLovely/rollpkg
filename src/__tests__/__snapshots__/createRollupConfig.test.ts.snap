// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`rollup plugins default build plugins with sideEffects: false 1`] = `
Array [
  "@rollup/plugin-resolve",
  "@rollup/plugin-commonjs",
  "@rollup/plugin-json",
  Object {
    "calledWith": Object {
      "include": Array [
        "**/*.ts+(|x)",
        "**/*.js+(|x)",
      ],
      "tsconfig": undefined,
      "tsconfigDefaults": Object {
        "compilerOptions": Object {
          "declaration": true,
          "rootDir": "./src",
          "types": Array [],
        },
        "include": Array [
          "src",
        ],
      },
      "tsconfigOverride": Object {
        "compilerOptions": Object {
          "module": "esnext",
          "sourceMap": true,
        },
        "exclude": Array [
          "**/__tests__",
          "**/__mocks__",
          "**/__snapshots__",
          "**/*.test.*",
          "**/*.spec.*",
          "**/*.mock.*",
        ],
        "include": Array [
          "./node_modules/rollpkg/configs/types",
        ],
      },
    },
    "name": "rollup-plugin-typescript2",
  },
  "rollup-plugin-sourcemaps",
  Object {
    "calledWith": Object {
      "__DEV__": "process.env.NODE_ENV !== 'production'",
    },
    "name": "@rollup/plugin-replace",
  },
]
`;

exports[`rollup plugins default build plugins with sideEffects: true 1`] = `
Array [
  "@rollup/plugin-resolve",
  "@rollup/plugin-commonjs",
  "@rollup/plugin-json",
  Object {
    "calledWith": Object {
      "include": Array [
        "**/*.ts+(|x)",
        "**/*.js+(|x)",
      ],
      "tsconfig": undefined,
      "tsconfigDefaults": Object {
        "compilerOptions": Object {
          "declaration": true,
          "rootDir": "./src",
          "types": Array [],
        },
        "include": Array [
          "src",
        ],
      },
      "tsconfigOverride": Object {
        "compilerOptions": Object {
          "module": "esnext",
          "sourceMap": true,
        },
        "exclude": Array [
          "**/__tests__",
          "**/__mocks__",
          "**/__snapshots__",
          "**/*.test.*",
          "**/*.spec.*",
          "**/*.mock.*",
        ],
        "include": Array [
          "./node_modules/rollpkg/configs/types",
        ],
      },
    },
    "name": "rollup-plugin-typescript2",
  },
  "rollup-plugin-sourcemaps",
  Object {
    "calledWith": Object {
      "__DEV__": "process.env.NODE_ENV !== 'production'",
    },
    "name": "@rollup/plugin-replace",
  },
]
`;

exports[`rollup plugins dev build plugins with sideEffects: false 1`] = `
Array [
  "@rollup/plugin-resolve",
  "@rollup/plugin-commonjs",
  "@rollup/plugin-json",
  Object {
    "calledWith": Object {
      "include": Array [
        "**/*.ts+(|x)",
        "**/*.js+(|x)",
      ],
      "tsconfig": undefined,
      "tsconfigDefaults": Object {
        "compilerOptions": Object {
          "declaration": true,
          "rootDir": "./src",
          "types": Array [],
        },
        "include": Array [
          "src",
        ],
      },
      "tsconfigOverride": Object {
        "compilerOptions": Object {
          "module": "esnext",
          "sourceMap": true,
        },
        "exclude": Array [
          "**/__tests__",
          "**/__mocks__",
          "**/__snapshots__",
          "**/*.test.*",
          "**/*.spec.*",
          "**/*.mock.*",
        ],
        "include": Array [
          "./node_modules/rollpkg/configs/types",
        ],
      },
    },
    "name": "rollup-plugin-typescript2",
  },
  "rollup-plugin-sourcemaps",
  Object {
    "calledWith": Object {
      "__DEV__": "process.env.NODE_ENV !== 'production'",
    },
    "name": "@rollup/plugin-replace",
  },
  Object {
    "calledWith": Object {
      "process.env.NODE_ENV": "\\"development\\"",
    },
    "name": "@rollup/plugin-replace",
  },
]
`;

exports[`rollup plugins dev build plugins with sideEffects: true 1`] = `
Array [
  "@rollup/plugin-resolve",
  "@rollup/plugin-commonjs",
  "@rollup/plugin-json",
  Object {
    "calledWith": Object {
      "include": Array [
        "**/*.ts+(|x)",
        "**/*.js+(|x)",
      ],
      "tsconfig": undefined,
      "tsconfigDefaults": Object {
        "compilerOptions": Object {
          "declaration": true,
          "rootDir": "./src",
          "types": Array [],
        },
        "include": Array [
          "src",
        ],
      },
      "tsconfigOverride": Object {
        "compilerOptions": Object {
          "module": "esnext",
          "sourceMap": true,
        },
        "exclude": Array [
          "**/__tests__",
          "**/__mocks__",
          "**/__snapshots__",
          "**/*.test.*",
          "**/*.spec.*",
          "**/*.mock.*",
        ],
        "include": Array [
          "./node_modules/rollpkg/configs/types",
        ],
      },
    },
    "name": "rollup-plugin-typescript2",
  },
  "rollup-plugin-sourcemaps",
  Object {
    "calledWith": Object {
      "__DEV__": "process.env.NODE_ENV !== 'production'",
    },
    "name": "@rollup/plugin-replace",
  },
  Object {
    "calledWith": Object {
      "process.env.NODE_ENV": "\\"development\\"",
    },
    "name": "@rollup/plugin-replace",
  },
]
`;

exports[`rollup plugins output default plugins 1`] = `Array []`;

exports[`rollup plugins output prod plugins 1`] = `
Array [
  Object {
    "calledWith": Object {
      "format": Object {
        "comments": false,
      },
    },
    "name": "rollup-plugin-terser",
  },
]
`;

exports[`rollup plugins prod build plugins with sideEffects: false 1`] = `
Array [
  "@rollup/plugin-resolve",
  "@rollup/plugin-commonjs",
  "@rollup/plugin-json",
  Object {
    "calledWith": Object {
      "include": Array [
        "**/*.ts+(|x)",
        "**/*.js+(|x)",
      ],
      "tsconfig": undefined,
      "tsconfigDefaults": Object {
        "compilerOptions": Object {
          "declaration": true,
          "rootDir": "./src",
          "types": Array [],
        },
        "include": Array [
          "src",
        ],
      },
      "tsconfigOverride": Object {
        "compilerOptions": Object {
          "module": "esnext",
          "sourceMap": true,
        },
        "exclude": Array [
          "**/__tests__",
          "**/__mocks__",
          "**/__snapshots__",
          "**/*.test.*",
          "**/*.spec.*",
          "**/*.mock.*",
        ],
        "include": Array [
          "./node_modules/rollpkg/configs/types",
        ],
      },
    },
    "name": "rollup-plugin-typescript2",
  },
  "rollup-plugin-sourcemaps",
  Object {
    "calledWith": Object {
      "__DEV__": "process.env.NODE_ENV !== 'production'",
    },
    "name": "@rollup/plugin-replace",
  },
  Object {
    "calledWith": Object {
      "process.env.NODE_ENV": "\\"production\\"",
    },
    "name": "@rollup/plugin-replace",
  },
]
`;

exports[`rollup plugins prod build plugins with sideEffects: true 1`] = `
Array [
  "@rollup/plugin-resolve",
  "@rollup/plugin-commonjs",
  "@rollup/plugin-json",
  Object {
    "calledWith": Object {
      "include": Array [
        "**/*.ts+(|x)",
        "**/*.js+(|x)",
      ],
      "tsconfig": undefined,
      "tsconfigDefaults": Object {
        "compilerOptions": Object {
          "declaration": true,
          "rootDir": "./src",
          "types": Array [],
        },
        "include": Array [
          "src",
        ],
      },
      "tsconfigOverride": Object {
        "compilerOptions": Object {
          "module": "esnext",
          "sourceMap": true,
        },
        "exclude": Array [
          "**/__tests__",
          "**/__mocks__",
          "**/__snapshots__",
          "**/*.test.*",
          "**/*.spec.*",
          "**/*.mock.*",
        ],
        "include": Array [
          "./node_modules/rollpkg/configs/types",
        ],
      },
    },
    "name": "rollup-plugin-typescript2",
  },
  "rollup-plugin-sourcemaps",
  Object {
    "calledWith": Object {
      "__DEV__": "process.env.NODE_ENV !== 'production'",
    },
    "name": "@rollup/plugin-replace",
  },
  Object {
    "calledWith": Object {
      "process.env.NODE_ENV": "\\"production\\"",
    },
    "name": "@rollup/plugin-replace",
  },
]
`;
